// seon-lib-implementation/dist/prefabs/sintaxes/typescript/TSTypes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypescriptPromisseAny = exports.TypescriptBigInt = exports.TypescriptSymbol = exports.TypescriptUndefined = exports.TypescriptNull = exports.TypescriptBoolean = exports.TypescriptNumber = exports.TypescriptString = void 0;
const PrimitiveTypeAbstraction_1 = __importDefault(require("../../../abstractions/types/PrimitiveTypeAbstraction"));
class TypescriptString extends PrimitiveTypeAbstraction_1.default {
    constructor() {
        super("string");
    }
}
exports.TypescriptString = TypescriptString;
class TypescriptNumber extends PrimitiveTypeAbstraction_1.default {
    constructor() {
        super("number");
    }
}
exports.TypescriptNumber = TypescriptNumber;
class TypescriptBoolean extends PrimitiveTypeAbstraction_1.default {
    constructor() {
        super("boolean");
    }
}
exports.TypescriptBoolean = TypescriptBoolean;
class TypescriptNull extends PrimitiveTypeAbstraction_1.default {
    constructor() {
        super("null");
    }
}
exports.TypescriptNull = TypescriptNull;
class TypescriptUndefined extends PrimitiveTypeAbstraction_1.default {
    constructor() {
        super("undefined");
    }
}
exports.TypescriptUndefined = TypescriptUndefined;
class TypescriptSymbol extends PrimitiveTypeAbstraction_1.default {
    constructor() {
        super("symbol");
    }
}
exports.TypescriptSymbol = TypescriptSymbol;
class TypescriptBigInt extends PrimitiveTypeAbstraction_1.default {
    constructor() {
        super("bigint");
    }
}
exports.TypescriptBigInt = TypescriptBigInt;
class TypescriptPromisseAny extends PrimitiveTypeAbstraction_1.default {
    constructor() {
        super("Promise<any>");
    }
}
exports.TypescriptPromisseAny = TypescriptPromisseAny;
