// seon-lib-implementation/dist/prefabs/sintaxes/Cpp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const Sintaxe_1 = __importStar(require("../abstractions/language/Sintaxe"));
const Enums_1 = require("../abstractions/shared/Enums");
class CppClasImplementation {
    classKeyword(clazz) {
        return "class";
    }
    className(clazz) {
        return clazz.getName();
    }
    classVisibility(clazz) {
        return "";
    }
    classModifier(clazz) {
        return "";
    }
    attrVisibility(attr) {
        switch (attr.getVisibility()) {
            case Enums_1.VisibilityOptions.PRIVATE:
                return "private: ";
        }
        return "";
    }
    attrModifier(attr) {
        return "";
    }
    attrName(attr) {
        return attr.getName();
    }
    attrType(attr) {
        return attr.getType().getName();
    }
    constructorVisibility(clazz) {
        return "";
    }
    constructorKeyword(clazz) {
        return clazz.getName();
    }
    argumentName(arg) {
        return arg.getName();
    }
    argumentType(arg) {
        return arg.getType().getName();
    }
    methodKeyword(method) {
        return "";
    }
    methodVisibility(method) {
        switch (method.getVisibility()) {
            case Enums_1.VisibilityOptions.PRIVATE:
                return "private: ";
        }
        return "public: ";
    }
    methodModifier(method) {
        switch (method.getOption()) {
            case Enums_1.ModifierOptions.STATIC:
                return "static";
        }
        return "";
    }
    methodName(method) {
        return method.getName();
    }
    methodReturn(method) {
        return method.getReturn().getName();
    }
}
class CppClassImplementationOrder extends Sintaxe_1.ClassImplementationOrder {
    getDeclarationOrder() {
        if (this.clazz == undefined)
            return [];
        return [
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.classKeyword),
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.className, ""),
        ];
    }
    getAttributeDeclarationOrder() {
        if (this.clazz == undefined)
            return [];
        return [
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.attrVisibility),
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.attrModifier),
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.attrType),
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.attrName),
        ];
    }
    getConstructorDeclarationOrder() {
        if (this.clazz == undefined)
            return [];
        return [
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.constructorKeyword),
        ];
    }
    getArggumentDeclarationOrder() {
        if (this.clazz == undefined)
            return [];
        return [
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.argumentType),
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.argumentName, ""),
        ];
    }
    getMethodsDeclarationOrder() {
        if (this.clazz == undefined)
            return [];
        return [
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.methodVisibility),
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.methodModifier),
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.methodReturn),
            new Sintaxe_1.ClassImplementationConfigOrder(this.clazz.methodName),
        ];
    }
    getMehodEmptyBody() {
        return "";
    }
}
const cppReservedData = {
    classReference: "this",
    classSubAccess: ".",
    endLineCode: ";",
    endCodeBlock: "}",
    startCodeBlock: "{",
    separeArgDeclaraion: ",",
    startArgDeclaration: "(",
    endArgDeclaraion: ")"
};
const cppExtensionFiles = {
    main: "cpp",
    others: ["h", "hpp", "c++"]
};
class CppSintaxe extends Sintaxe_1.default {
    constructor() {
        super(new CppClassImplementationOrder(), cppReservedData, cppExtensionFiles);
    }
}
exports.default = CppSintaxe;
