// seon-lib-implementation/dist/prefabs/sintaxes/Python.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PythonString = exports.PythonInteger = void 0;
const ClassAbstraction_js_1 = __importDefault(require("../abstractions/oo/ClassAbstraction.js"));
const PrimitiveTypeAbstraction_js_1 = __importDefault(require("../abstractions/types/PrimitiveTypeAbstraction.js"));
const Sintaxe_js_1 = __importStar(require("../abstractions/language/Sintaxe.js"));
const Enums_js_1 = require("../abstractions/shared/Enums.js");
const pythonReservedData = {
    startCodeBlock: ":",
    endCodeBlock: "",
    endLineCode: "",
    startArgDeclaration: "(",
    endArgDeclaraion: ")",
    separeArgDeclaraion: ", ",
    classReference: "self",
    classSubAccess: ".",
};
const pythonExtensionFile = {
    main: "py",
};
const pythonABC = new ClassAbstraction_js_1.default("ABC", [], []);
class PythonInteger extends PrimitiveTypeAbstraction_js_1.default {
    constructor() {
        super("int");
    }
}
exports.PythonInteger = PythonInteger;
class PythonString extends PrimitiveTypeAbstraction_js_1.default {
    constructor() {
        super("str");
    }
}
exports.PythonString = PythonString;
class PythonVisibilityWraper {
    wrapVisibility(visibility) {
        switch (visibility) {
            case Enums_js_1.VisibilityOptions.PUBLIC:
                return "";
            case Enums_js_1.VisibilityOptions.PROTECTED:
                return "_";
            case Enums_js_1.VisibilityOptions.PRIVATE:
                return "__";
        }
    }
}
const pythonVWraper = new PythonVisibilityWraper();
class PythonClassImplementation {
    classKeyword(clazz) {
        return "class";
    }
    className(clazz) {
        return clazz.getName();
    }
    classVisibility(clazz) {
        return pythonVWraper.wrapVisibility(clazz.getVisibility());
    }
    classModifier(clazz) {
        switch (clazz.getModifier()) {
            case Enums_js_1.ModifierOptions.ABSTRACT:
                clazz.addHerance(pythonABC);
        }
        return "";
    }
    attrVisibility(attr) {
        return pythonVWraper.wrapVisibility(attr.getVisibility());
    }
    attrModifier(attr) {
        return "";
    }
    attrName(attr) {
        return attr.getName();
    }
    attrType(attr) {
        return attr.getType().getName();
    }
    constructorVisibility(clazz) {
        return "";
    }
    constructorKeyword(clazz) {
        return "def __init__";
    }
    argumentName(arg) {
        return arg.getName();
    }
    argumentType(arg) {
        return arg.getType().getName();
    }
    methodKeyword(method) {
        return "def";
    }
    methodVisibility(method) {
        return pythonVWraper.wrapVisibility(method.getVisibility());
    }
    methodModifier(method) {
        return "";
    }
    methodName(method) {
        return method.getName();
    }
    methodReturn(method) {
        return `-> ${method.getReturn()}`;
    }
}
class PythonImplementationOrder extends Sintaxe_js_1.ClassImplementationOrder {
    constructor() {
        super(new PythonClassImplementation());
    }
    getDeclarationOrder() {
        if (this.clazz == undefined)
            return [];
        return [
            new Sintaxe_js_1.ClassImplementationConfigOrder(this.clazz.classKeyword),
            new Sintaxe_js_1.ClassImplementationConfigOrder(this.clazz.classVisibility, ""),
            new Sintaxe_js_1.ClassImplementationConfigOrder(this.clazz.className, ""),
        ];
    }
    getAttributeDeclarationOrder() {
        return [];
    }
    getConstructorDeclarationOrder() {
        if (this.clazz == undefined)
            return [];
        return [
            new Sintaxe_js_1.ClassImplementationConfigOrder(this.clazz.constructorKeyword),
        ];
    }
    getArggumentDeclarationOrder() {
        if (this.clazz == undefined)
            return [];
        return [
            new Sintaxe_js_1.ClassImplementationConfigOrder(this.clazz.argumentName, ": "),
            new Sintaxe_js_1.ClassImplementationConfigOrder(this.clazz.argumentType, ""),
        ];
    }
    getMethodsDeclarationOrder() {
        if (this.clazz == undefined)
            return [];
        return [
            new Sintaxe_js_1.ClassImplementationConfigOrder(this.clazz.methodKeyword),
            new Sintaxe_js_1.ClassImplementationConfigOrder(this.clazz.methodVisibility),
            new Sintaxe_js_1.ClassImplementationConfigOrder(this.clazz.methodName),
            new Sintaxe_js_1.ClassImplementationConfigOrder((obj) => { return "->"; }),
            new Sintaxe_js_1.ClassImplementationConfigOrder(this.clazz.methodReturn),
        ];
    }
    getMehodEmptyBody() {
        return "pass";
    }
}
class PythonSintaxe extends Sintaxe_js_1.default {
    constructor() {
        super(new PythonImplementationOrder(), pythonReservedData, pythonExtensionFile);
    }
}
exports.default = PythonSintaxe;
